idx/memory/tag_query.go

```diff
@@ -49,37 +71,66 @@ func (a KvReByCost) Len() int           { return len(a) }
 func (a KvReByCost) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a KvReByCost) Less(i, j int) bool { return a[i].cost < a[j].cost }
 
+//TagQuery runs a set of pattern or string matches on tag keys and values against
+// the index. It is executed via:
+// Run() which returns a set of matching MetricIDs
+// RunGetTags() which returns the set of tag values found amongst the matching MetricIDs
 type TagQuery struct {
-	from      int64
-	equal     []kv
-	match     []kvRe
-	notEqual  []kv
-	notMatch  []kvRe
-	startWith int
+	from        int64
+	equal       []kv
+	match       []kvRe
+	notEqual    []kv
+	notMatch    []kvRe
+	prefix      []kv
+	startWith   int // expression type to generate the initial result set from
+	filterTagBy int // if tags should be filtered by prefix or regex match (max one)
+
+	// no need have more than one of tagMatch and tagPrefix, tags can only be
+	// filtered by max one condition
+	tagMatch  kvRe   // only used for /metrics/tags with regex in filter param
+	tagPrefix string // only used for auto complete of tags to match exact prefix
```

Comment:
it's not quite clear why we need these attributes since we already have all the conditions above (match, notMatch, notEqual, etc). we need to document that the above are for expressions to query the index, but what are these two for?

Reply:
it's true that we have all the conditions above. but since we support this special tag `__tag` which makes the condition match against the tag itself, instead of matching against the value, those are internally separated into the `tagMatch` and `tagPrefix` attribute. this makes sense because the logic of the expression evaluation to match against the tag is quite different from matching against the value.
as the comment says, there is no need to make `tagMatch`/`tagPrefix` a list, because we only allow one condition against the `__tag`

Reply:
https://github.com/grafana/metrictank/pull/779/commits/ae37dc480178902c3759c93a125f517e7eccbab0